log4j-core/src/main/java/org/apache/logging/log4j/core/util/internal/InternalLoggerRegistry.java

```diff
@@ -147,44 +147,43 @@ public Logger computeIfAbsent(
+                                    + "Please report this using the Log4j project issue tracker.",
+                            loggerName,
+                            loggerMessageFactory,
+                            name,
+                            messageFactory);
+        }
+
         // Write lock slow path: Insert the logger
         writeLock.lock();
         try {
 
-            // See if the logger is created by another thread in the meantime
-            final Map<String, WeakReference<Logger>> loggerRefByName =
-                    loggerRefByNameByMessageFactory.computeIfAbsent(messageFactory, ignored -> new HashMap<>());
-            WeakReference<Logger> loggerRef = loggerRefByName.get(name);
-            if (loggerRef != null && (logger = loggerRef.get()) != null) {
-                return logger;
-            }
-
-            // Create the logger
-            logger = loggerSupplier.apply(name, messageFactory);
-
-            // Report name and message factory mismatch if there are any
-            final String loggerName = logger.getName();
-            final MessageFactory loggerMessageFactory = logger.getMessageFactory();
-            if (!loggerMessageFactory.equals(messageFactory)) {
-                StatusLogger.getLogger()
-                        .error(
-                                "Newly registered logger with name `{}` and message factory `{}`, is requested to be associated with a different name `{}` or message factory `{}`.\n"
-                                        + "Effectively the message factory of the logger will be used and the other one will be ignored.\n"
-                                        + "This generally hints a problem at the logger context implementation.\n"
-                                        + "Please report this using the Log4j project issue tracker.",
-                                loggerName,
-                                loggerMessageFactory,
-                                name,
-                                messageFactory);
-                // Register logger under alternative keys
-                loggerRefByNameByMessageFactory
-                        .computeIfAbsent(loggerMessageFactory, ignored -> new HashMap<>())
-                        .putIfAbsent(loggerName, new WeakReference<>(logger));
-            }
-
-            // Insert the logger
-            loggerRefByName.put(name, new WeakReference<>(logger));
-            return logger;
+            Logger currentLogger = loggerRefByNameByMessageFactory
+                    .computeIfAbsent(messageFactory, ignored -> new HashMap<>())
+                    .computeIfAbsent(name, ignored -> new WeakReference<>(newLogger))
+                    .get();
+            // A replacement for Reference.reachabilityFence() from Java 9.
+            // Prevents `newLogger` to become unreachable in the lines above.
+            return currentLogger != null ? currentLogger : newLogger;
```

Comment:
Are you sure this is necessary? If `WR` returns _a non-null value_, and you hold a reference to that, I don't think `WR` can be emptied in the meantime. Am I mistaken?

Reply:
You might be right, but I am not eager to risk another `2.24.1` :smile: 

My reasoning was:

- We get a `WR` to `newLogger`.
- At that point in code `newLogger` has already been referenced for the last time. The `() -> new WeakReference<>(newLogger)` has already been used and can be reclaimed.
- If a GC occurs just before the `WR.get()`, the reference might be emptied.

Since I reference `newLogger` on line 186, `currentLogger` will never be `null`.

Reply:
> At that point in code `newLogger` has already been referenced for the last time. The `() -> new WeakReference<>(newLogger)` has already been used and can be reclaimed.

No, it cannot; because there is still a strong reference to the instance.

> You might be right, but I am not eager to risk another `2.24.1` 😄

Could you at least document that this is a _"speculative measure"_?

Reply:
Removed this in 14c6df0d8053e4f312bf72cd117ce8488f0a43df – even IntelliJ warned that this is redundant.